### DISCUSSION TITLE
How to fix invalid chunks,

### DISCUSSION THREAD

**Discussion by kiranranjane** [moosefs/moosefs#383] [Q&A]
I am getting lots of invalid chunks after a power outage and due to this virtual machines go in read-only mode or I have to power off and power on it back to make it working again, this is keep on going now, How do I fix this, All production VMs are getting affected.

Is there a procedure to fix this, Moosefs version 113 with xfs filesystem, they are separate physical nodes only used for storage. And this issue is only on 2 chunk servers, rest are fine. Total 9 chunkservers however these 2 are the one with most drives and storage capacity.

Please assist.

**Comment by pkonopelko** [comment]
In such case you may mostly have "invalid copies" since chunk's header that contains CRC checksums is written at the end of chunk's writing process. If you had a power outage, most probably chunks' headers, versions and CRC sums may be corrupted – especially for chunks that data was being written to when the outage occurred. However, usually in such cases such chunks' contents is fine.

1. If your "filesystem check info" loop has already passed one time, in the "Info" tab in MFSCGI you should be able to see the list of currently unavailable files and chunks an types of missing chunks

2. Use `mfsfileinfo` on unavailable files to check on what chunkservers their invalid / wrong version chunks are stored (e.g. `mfsfileinfo file | grep -A3 -B3 INVALID` or `mfsfileinfo file | grep -A3 -B3 WRONG`).

3. If you have the same chunk's copies with different missing chunk types (e.g. one copy is "wrong version" and 2nd is "invalid", please check their `ctime`s / `mtime`s. Usually the newer is "better" and usually (considering VMs case) "invalid copies" are "better" than "wrong versions". But each case is different and it is tough to say something concrete without an access and seeing the situation...

3. For "invalid copies" you need to manually fix the chunks on the chunkserver. **Stop the Chunkserver's process first**, find a chunk (like `find /mnt/mfschunksdisk* -type f -iname "*XXXXX*"`, where `XXXXX` is a non-zeroes part of the chunk's number) on the chunkservers and use `mfschunktool -x chunk_name.mfs` in order to check it. Then **make a backup of the chunk** and use `mfschunktool -rx chunk_name.mfs` in order to fix the chunk's header (CRC, version). For more information refer to `man mfschunktool`. Then remove `.chunkdb` file on the disk that you were fixing a chunk on and start the Chunkserver process.

3. After dealing with "invalid copies", if you have "wrong versions", you can use `mfsfilerepair` for such type of missing chunks. Note that it works on **files** (so `mfsmount`'s side), **not chunks**. Please read `man mfsfilerepair` first, I'm pasting its important excerpt below:

> In case of missing chunk it fills missing parts of file with zeros; in case of chunk version mismatch it sets chunk version known to mfsmaster to highest one found on chunkservers. Note: because in the second case content mismatch can occur in chunks with the same version, it is  advised  to make a copy (not a snapshot!) and delete original file after "repairing".

6. In case of "no copy" you may want to look for the files on disks considered as damaged first. If you are sure that "no copy" chunks are completely lost, you can use `mfsfilerepair`, but please remember that in such case it will make the file (partially) readable by **filling missing (no copy) chunks with zeroes**, so make double sure you really want this.

7. After file (VM disk image) recovery is completed, it is recommended to mount the image and run `fsck` against it, prior to running the VM's OS for the disk image.

This is a long process unfortunately, but worth going through. Hope it will help you.

**Comment by pkonopelko** [comment]
PS: It is also good to make a snapshot (e.g.: `mfsmakesnapshot /mnt/mfs/VM_disk_image.img VM_disk_image_SNAP.img`) of a corrupted file before starting any action, just in case.
